File: benchmarks/Ruby/permute.rb
# frozen_string_literal: true

# This code is derived from the SOM benchmarks, see AUTHORS.md file.
#
# Copyright (c) 2015-2016 Stefan Marr <git@stefan-marr.de>
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the 'Software'), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in
# all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED 'AS IS', WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
# THE SOFTWARE.

require_relative 'benchmark'

class Permute < Benchmark
  def initialize
    @count = 0
    @v     = nil
  end

  def benchmark
    @count = 0
    @v = Array.new(6, 0)
    permute(6)
    @count
  end

  def permute(n)
    @count += 1
    if n != 0
      n1 = n - 1
      permute(n1)

      n1.downto(0) do |i|
        swap(n1, i)
        permute(n1)
        swap(n1, i)
      end
    end
  end

  def swap(i, j)
    tmp = @v[i]
    @v[i] = @v[j]
    @v[j] = tmp
  end

  def verify_result(result)
    result == 8660
  end
end

File: benchmarks/Rust/src/permute.rs
use crate::benchmark::Benchmark;
use std::any::Any;

#[derive(Default)]
pub struct Permute {
    count: usize,
    v: [usize; 6],
}

impl Benchmark for Permute {
    fn benchmark(&self) -> Box<dyn Any> {
        let mut permute = Permute::default();
        permute.permute(6);
        Box::new(permute.count)
    }

    fn verify_result(&self, result: Box<dyn Any>) -> bool {
        if let Ok(result) = result.downcast::<usize>() {
            *result == 8660
        } else {
            false
        }
    }
}

impl Permute {
    fn permute(&mut self, n: usize) {
        self.count += 1;
        if n != 0 {
            let n1 = n - 1;
            self.permute(n1);
            for i in (0..=n1).rev() {
                self.swap(n1, i);
                self.permute(n1);
                self.swap(n1, i);
            }
        }
    }

    fn swap(&mut self, i: usize, j: usize) {
        self.v.swap(i, j);
    }
}
